src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowLoopSpec.groovy

```diff
@@ -0,0 +1,618 @@
+
+        when: "Create flowLoop on the sw switch"
+        def createResponse = northboundV2.createFlowLoop(flow.flowId, new FlowLoopPayload(sw.dpId))
+
+        then: "Create flowLoop response contains flowId and src switchId"
+        assert createResponse.flowId == flow.flowId
+        assert createResponse.switchId == sw.dpId
+
+        and: "Flow is UP and valid"
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.flowId).status == FlowState.UP }
+        northbound.validateFlow(flow.flowId).each { direction -> assert direction.asExpected }
+
+        and: "FlowLoop is really created on the switch"
+        with(northbound.getFlow(flow.flowId)) {
+            it.looped
+            it.loopSwitchId == sw.dpId
+        }
+
+        and: "FlowLoop rules are created"
+        //bug
+//    Wrappers.wait(RULES_INSTALLATION_TIME) {
+//        assert northbound.getSwitchRules(sw.dpId).flowEntries.findAll {
+//            new Cookie(it.cookie).getType() == CookieType.FLOW_LOOP
+//        }.size() == 2
+//    }
+
+        and: "The switch is valid"
+        //there is a bug, flow_loop is in excess
+//      northbound.validateSwitch(sw.dpId).verifyRuleSectionsAreEmpty(["missing", "excess", "misconfigured"])
+
+        when: "Delete the flow with created flowLoop"
+        flowHelperV2.deleteFlow(flow.flowId)
+        def flowIsDeleted = true
+
+        then: "FlowLoop rules are deleted on the switch"
+        assert northbound.getSwitchRules(sw.dpId).flowEntries.findAll {
+            new Cookie(it.cookie).getType() == CookieType.FLOW_LOOP
+        }.empty
+
+        and: "The switch is valid"
+        northbound.validateSwitch(sw.dpId).verifyRuleSectionsAreEmpty(["missing", "excess", "misconfigured"])
+        def testIsCompleted = true
+
+        cleanup: "Delete the flow"
+        !flowIsDeleted && flowHelperV2.deleteFlow(flow.flowId)
+        !testIsCompleted && northbound.synchronizeSwitch(sw.dpId, true)
+    }
+
+    @Tidy
+    def "System is able to reroute a flow when flowLoop is created on it"() {
+        given: "A multi switch flow with one alternative path at least"
+        def switchPair = topologyHelper.getAllNeighboringSwitchPairs().find {
+            it.paths.unique(false) {
+                a, b -> a.intersect(b) == [] ? 1 : 0
+            }.size() >= 2
+        }
+        def flow = flowHelperV2.randomFlow(switchPair)
+        flowHelperV2.addFlow(flow)
+        def flowPath = PathHelper.convert(northbound.getFlowPath(flow.flowId))
+
+        and: "FlowLoop is created on the dst switch"
+        northboundV2.createFlowLoop(flow.flowId, new FlowLoopPayload(switchPair.dst.dpId))
+
+        when: "Fail a flow ISL (bring switch port down)"
+        def allFlowPaths = switchPair.paths
+        Set<Isl> altFlowIsls = []
+        def flowIsls = pathHelper.getInvolvedIsls(flowPath)
+        allFlowPaths.findAll { it != flowPath }.each { altFlowIsls.addAll(pathHelper.getInvolvedIsls(it)) }
```

Comment:
this is 100% fine code. Just a hint, you can make it 1 line
`Set<Isl> altFlowIsls = allFlowPaths.findAll{ it != flowPath }.collectMany{ pathHelper.getInvolvedIsls(it) }`